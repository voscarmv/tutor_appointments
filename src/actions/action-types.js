export const FETCH_BREEDS_REQUEST = 'FETCH_BREEDS_REQUEST';
export const FETCH_BREEDS_SUCCESS = 'FETCH_BREEDS_SUCCESS';
export const FETCH_BREEDS_ERROR = 'FETCH_BREEDS_ERROR';
export const FETCH_CAT_REQUEST = 'FETCH_CAT_REQUEST';
export const FETCH_CAT_SUCCESS = 'FETCH_CAT_SUCCESS';
export const FETCH_CAT_ERROR = 'FETCH_CAT_ERROR';
export const FETCH_SIGNUP_REQUEST = 'FETCH_SIGNUP_REQUEST';
export const FETCH_SIGNUP_SUCCESS = 'FETCH_SIGNUP_SUCCESS';
export const FETCH_SIGNUP_ERROR = 'FETCH_SIGNUP_ERROR';
export const FETCH_LOGIN_REQUEST = 'FETCH_LOGIN_REQUEST';
export const FETCH_LOGIN_SUCCESS = 'FETCH_LOGIN_SUCCESS';
export const FETCH_LOGIN_ERROR = 'FETCH_LOGIN_ERROR';
export const LOGIN_DISMISS = 'LOGIN_DISMISS';
export const UPDATE_FILTER = 'UPDATE_FILTER';
export const ALERT_MESSAGE = 'ALERT_MESSAGE';
export const AUTH_KEY = 'AUTH_KEY';
