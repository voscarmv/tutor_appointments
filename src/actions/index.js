import fetch from 'node-fetch';
import {
  DISMISS,
  DISMISS_SUBJECT,
  ALERT_MESSAGE,
  AUTH_KEY,
  AUTH_CLEAR,
  UPDATE_SUBJECT,
  DISMISS_APPOINTMENT,
  FETCH_REQUEST,
  FETCH_SUCCESS,
  FETCH_ERROR,
} from './action-types';

export const alertMessage = payload => ({ type: ALERT_MESSAGE, payload });
export const authKey = payload => ({ type: AUTH_KEY, payload });
export const authClear = payload => ({ type: AUTH_CLEAR, payload });

export const fetchRequest = message => ({
  type: FETCH_REQUEST,
  message,
});

export const fetchSuccess = (data, message) => ({
  type: FETCH_SUCCESS,
  payload: data,
  message,
});

export const fetchError = (error, message) => ({
  type: FETCH_ERROR,
  payload: error,
  message,
});

export const fetchSignUpRequest = () => fetchRequest({
  content: 'Signing up...',
  type: 'info',
  show: true,
});
export const fetchSignUpSuccess = data => (fetchSuccess(data, {
  content: 'Signed up successfully!',
  type: 'success',
  show: true,
}));
export const fetchSignUpError = error => (fetchError(error, {
  content: `Failed to sign up: ${error}`,
  type: 'danger',
  show: true,
}));

export const dismissAlert = () => ({ type: DISMISS });

export const fetchLogInRequest = () => fetchRequest({
  content: 'Logging in...',
  type: 'info',
  show: true,
});
export const fetchLogInSuccess = data => (fetchSuccess(data, {
  content: 'Logged in successfully!',
  type: 'success',
  show: true,
}));
export const fetchLogInError = error => (fetchError(error, {
  content: `Failed to log in: ${error}`,
  type: 'danger',
  show: true,
}));

export const fetchLogIn = (data, history) => async dispatch => {
  dispatch(fetchLogInRequest());
  try {
    const jsonUpdate = data;
    const getLogin = await fetch(
      'https://thawing-beyond-27762.herokuapp.com/login',
      {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
        },
        body: JSON.stringify(jsonUpdate),
      },
    );
    const loginJSON = await getLogin.json();

    const userID = loginJSON.id;
    const userEmail = loginJSON.email;
    const loginJWT = loginJSON.key;
    if (loginJWT === undefined) {
      throw new Error('Check your username and/or password.');
    }

    if (getLogin.status !== 200) {
      throw getLogin.statusText;
    }
    dispatch(authKey({ uid: userID, email: userEmail, key: loginJWT }));
    dispatch(fetchLogInSuccess(getLogin));
    history.push('/subjects');
  } catch (e) {
    dispatch(fetchLogInError(e));
  }
};

export const fetchSignUp = (data, history) => async dispatch => {
  dispatch(fetchSignUpRequest());
  try {
    const jsonUpdate = data;

    const getSignUp = await fetch(
      'https://thawing-beyond-27762.herokuapp.com/signup',
      {
        method: 'POST',
        mode: 'cors',
        headers: {
          'Content-type': 'application/json',
        },
        body: JSON.stringify(jsonUpdate),
      },
    );
    const signupJSON = await getSignUp.json();

    if (signupJSON.exception) {
      throw signupJSON.exception;
    }
    dispatch(fetchSignUpSuccess(signupJSON));
    dispatch(fetchLogIn(
      {
        email: data.email,
        password: data.password,
      }, history,
    ));
  } catch (e) {
    dispatch(fetchSignUpError(e));
  }
};

export const fetchLogOutRequest = () => fetchRequest({
  content: 'Logging out...',
  type: 'info',
  show: true,
});
export const fetchLogOutSuccess = data => (fetchSuccess(data, {
  content: 'Logged out successfully!',
  type: 'success',
  show: true,
}));
export const fetchLogOutError = error => (fetchError(error, {
  content: `Failed to log out: ${error}`,
  type: 'danger',
  show: true,
}));

export const fetchLogOut = (data, history) => async dispatch => {
  dispatch(dismissAlert());
  dispatch(fetchLogOutRequest());
  try {
    const getLogout = await fetch(
      'https://thawing-beyond-27762.herokuapp.com/logout',
      {
        method: 'DELETE',
        headers: {
          'content-type': 'application/json',
          Authorization: data,
        },
      },
    );
    if (getLogout.status !== 200) {
      throw getLogout.statusText;
    }
    dispatch(fetchLogOutSuccess());
    dispatch(authClear());
    history.push('/login');
  } catch (e) {
    dispatch(fetchLogOutError(e));
  }
};

export const dismissSubject = () => ({ type: DISMISS_SUBJECT });

export const fetchSubjectsRequest = () => fetchRequest({
  content: 'Fetching subjects...',
  type: 'info',
  show: false,
});
export const fetchSubjectsSuccess = data => (fetchSuccess(data, {
  content: 'Fetched subjects successfully!',
  type: 'success',
  show: false,
}));
export const fetchSubjectsError = error => (fetchError(error, {
  content: `Failed to fetch subjects: ${error}`,
  type: 'danger',
  show: false,
}));

export const fetchSubjects = data => async dispatch => {
  dispatch(fetchSubjectsRequest());
  try {
    const getSubjects = await fetch(
      'https://thawing-beyond-27762.herokuapp.com/subjects',
      {
        method: 'GET',
        headers: {
          'content-type': 'application/json',
          Authorization: data,
        },
      },
    );
    if (getSubjects.status !== 200) {
      throw getSubjects.statusText;
    }

    const subj = await getSubjects.json();
    dispatch(fetchSubjectsSuccess(subj));
  } catch (e) {
    dispatch(fetchSubjectsError(e));
  }
};

export const updateSubject = subject => ({ type: UPDATE_SUBJECT, payload: subject });

export const postAppointmentRequest = () => fetchRequest({
  content: 'Creating appointment...',
  type: 'info',
  show: false,
});
export const postAppointmentSuccess = data => (fetchSuccess(data, {
  content: 'Created appointment successfully!',
  type: 'success',
  show: false,
}));
export const postAppointmentError = error => (fetchError(error, {
  content: `Failed to create appointment: ${error}`,
  type: 'danger',
  show: false,
}));

export const postAppointment = (data, key, history) => async dispatch => {
  dispatch(postAppointmentRequest());
  try {
    const jsonUpdate = { appointment: data };

    const getAppointment = await fetch(
      'https://thawing-beyond-27762.herokuapp.com/appointments',
      {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          Authorization: key,
        },
        body: JSON.stringify(jsonUpdate),
      },
    );
    if (getAppointment.status !== 201) {
      throw getAppointment.statusText;
    }
    dispatch(postAppointmentSuccess());
    history.push('/appointmentslist');
  } catch (e) {
    dispatch(postAppointmentError(e));
  }
};

export const dismissAppointment = () => ({ type: DISMISS_APPOINTMENT });

export const fetchAppointmentRequest = () => fetchRequest({
  content: 'Fetching appointments...',
  type: 'info',
  show: false,
});
export const fetchAppointmentSuccess = data => (fetchSuccess(data, {
  content: 'Fetched appointments successfully!',
  type: 'success',
  show: false,
}));
export const fetchAppointmentError = error => (fetchError(error, {
  content: `Failed to fetch appointments: ${error}`,
  type: 'danger',
  show: false,
}));

export const fetchAppointment = data => async dispatch => {
  dispatch(fetchAppointmentRequest());
  try {
    const getAppointment = await fetch(
      'https://thawing-beyond-27762.herokuapp.com/appointments',
      {
        method: 'GET',
        headers: {
          'content-type': 'application/json',
          Authorization: data,
        },
      },
    );

    if (getAppointment.status !== 200) {
      throw getAppointment.statusText;
    }

    const appointmentJSON = await getAppointment.json();
    dispatch(fetchAppointmentSuccess(appointmentJSON));
  } catch (e) {
    dispatch(fetchAppointmentError(e));
  }
};
