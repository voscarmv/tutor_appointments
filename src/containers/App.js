/* eslint-disable no-console */
import React from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import Nav from './Nav';
import LogInPage from './LogInPage';
import SignUpPage from './SignUpPage';
import Home from './Home';
import SubjectsPage from './SubjectsPage';
import Appointment from './Appointment';

const App = () => (
  <Router>
    <Nav />
    <Switch>
      <Route exact path="/">
        <Home />
      </Route>
      <Route path="/login">
        <LogInPage />
      </Route>
      <Route path="/signup">
        <SignUpPage />
      </Route>
      <Route path="/subjects">
        <SubjectsPage />
      </Route>
      <Route path="/appointment">
        <Appointment />
      </Route>
    </Switch>
  </Router>
);

export default App;
