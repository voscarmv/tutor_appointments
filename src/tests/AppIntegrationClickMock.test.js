import React from 'react';
import {
  render, screen, waitFor, fireEvent,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../store/index';
import App from '../containers/App';
import 'regenerator-runtime/runtime';
import * as actions from '../actions/index';

jest.mock('../actions/index');

describe('App', () => {
  beforeEach(() => {
    actions.fetchSignUp.mockImplementation(
      () => ({
        type: 'FETCH_SUCCESS',
        payload: { id: 1, email: 'oscar@gmail.com', key: 'test_jwt' },
        message: {
          content: 'Signed up successfully!',
          type: 'success',
          show: true,
        },
      }),
    );
    actions.dismissAlert.mockImplementation(
      () => ({ type: 'DISMISS' }),
    );
  });
  afterEach(() => {
    actions.fetchSignUp.mockRestore();
    actions.dismissAlert.mockRestore();
  });
  test('Successful signup', async () => {
    const fullApp = render(<Provider store={store}><App /></Provider>);
    const { getByTestId, asFragment } = fullApp;
    fireEvent.click(getByTestId('signup_menu'));
    await waitFor(
      () => {
        expect(screen.getByText('We will never share your email with anyone else.'));
      },
    );
    fireEvent.change(getByTestId('email'), 'oscar@mail.com');
    fireEvent.change(getByTestId('password'), 'oscar@mail.com');
    fireEvent.click(getByTestId('submit_signup'));
    await waitFor(
      () => {
        expect(screen.getByText('Signed up successfully!'));
      },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
