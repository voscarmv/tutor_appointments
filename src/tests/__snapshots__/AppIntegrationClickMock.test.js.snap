// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Successful signup 1`] = `
<DocumentFragment>
  <div
    class="container-fluid d-flex h-100 flex-column h-100"
  >
    <div
      class="row flex-fill d-flex justify-content-start"
    >
      <nav
        class="flex-column col-lg-2 col-12 navbar navbar-expand-lg navbar-light"
      >
        <span
          class="navbar-brand"
        >
          <a
            class="vespa-text"
            href="#/"
          >
            <div>
              Tutor
              <br />
              appointments
            </div>
          </a>
        </span>
        <div
          class="navbar-collapse collapse"
          id="responsive-navbar-nav"
        >
          <div
            class="flex-column"
          >
            <div
              class="flex-column mt-auto mb-auto navbar-nav"
            >
              <div
                class="m-auto a-hover w-100 text-center d-none nav-item"
              >
                <a
                  class="nav-link-custom"
                  href="#/subjects"
                >
                  <div
                    class="link-box"
                  >
                    SUBJECTS
                  </div>
                </a>
              </div>
              <div
                class="m-auto a-hover w-100 text-center d-none nav-item"
              >
                <a
                  class="nav-link-custom"
                  href="#/appointmentslist"
                >
                  <div
                    class="link-box"
                  >
                    APPOINTMENTS
                  </div>
                </a>
              </div>
              <div
                class="m-auto a-hover w-100 text-center nav-item"
              >
                <a
                  class="nav-link-custom"
                  href="#/login"
                >
                  <div
                    class="link-box"
                  >
                    LOGIN
                  </div>
                </a>
              </div>
              <div
                class="m-auto a-hover w-100 text-center nav-item"
              >
                <a
                  class="nav-link-custom"
                  data-testid="signup_menu"
                  href="#/signup"
                >
                  <div
                    class="link-box"
                  >
                    SIGNUP
                  </div>
                </a>
              </div>
            </div>
            <div
              class="mt-auto d-none"
            >
              <div
                class="flex-column text-center"
              >
                <div />
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Log out
                </button>
              </div>
            </div>
          </div>
        </div>
        <button
          aria-controls="responsive-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
      </nav>
      <div
        class="col-lg-10 col-12 h-100"
      >
        <div>
          <div
            class="fade alert alert-success show"
            role="alert"
          >
            <p>
              Signed up successfully!
            </p>
            <p>
              <button
                class="btn btn-primary"
                type="button"
              >
                Ok
              </button>
            </p>
          </div>
          <h1>
            Sign up
          </h1>
          <form
            class=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="formBasicEmail"
              >
                Email address
              </label>
              <input
                class="form-control"
                data-testid="email"
                id="formBasicEmail"
                placeholder="Enter email"
                type="email"
              />
              <small
                class="text-muted form-text"
              >
                We will never share your email with anyone else.
              </small>
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="formBasicPassword"
              >
                Password
              </label>
              <input
                class="form-control"
                data-testid="password"
                id="formBasicPassword"
                minlength="8"
                placeholder="Password"
                type="password"
              />
            </div>
            <button
              class="btn btn-primary"
              data-testid="submit_signup"
              type="submit"
            >
              Sign up
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
